Rewrite ReactDOMSelection to use fewer ranges

sophiebits opened the pull request:
We heard from Chrome engineers that creating too many Range objects slows down Chrome because it needs to keep track of all of them for the case that anchor/focus nodes get removed from the document. We can just implement this calculation without ranges anyway.

spicyj commented:
I think this logic is right. I tested it by hand and it seemed to match in all cases I tried (in the Notes composer especially). Works with backwards selections. Apparently "old IE" (not sure which IEs) uses this path for textareas so I should probably test that too.

cc @n8schloss

spicyj commented:
> Are we really doing all this work for every commit?

When the focus is inside a contenteditable, apparently we are.

spicyj commented on src/renderers/dom/shared/ReactDOMSelection.js:
```diff
@@ -39,59 +32,67 @@ function getModernOffsets(node) {
+  let start = null;
+  let end = null;
+
+  function traverse(node) {
```
ok ok fine, I wanted to be lazy but I can do that too

spicyj replied:
I think I need it this way because I need the stack of indices. Since I only need the index if node is one of [anchorNode, focusNode] I could technically do it in constant space but it would be a huge pain. Can we leave it like this?

sebmarkbage replied:
You could have two index variables. On for anchor and one for focus. You only increment the index if you're currently traversing anchor or focus parent.

We could leave the recursion but can we at least avoid the closure by passing anchorNode/focusNode/anchorOffset/focusOffset as arguments? There are two return values but we're already allocating for that so just return the object from the recursive function.

Also, isn't text in selection something we know the optimal solution for? :)

sebmarkbage commented:
Oh ok. It's not as bad as I thought. I thought we did this for non-content editable. Phew.

sophiebits commented:
Rewritten with no closures or recursion. I am not proud of the complexity but I think it hits all the requirements.

I included a fuzz test against a much simpler (slower?) implementation, added to our test suite. jsdom doesn't support Range objects, but I copied the fuzz test code into my browser and manually compared it against our old implementation https://gist.github.com/sophiebits/2e6d571f4f10f33b62ea138a6e9c265c; with 200,000 trials no differences were found.

Think this should be good.

sophiebits commented on src/renderers/dom/shared/ReactDOMSelection.js:
```diff
@@ -39,59 +32,116 @@ function getModernOffsets(node) {
+    while (true) {
+      if (
+        node === anchorNode &&
+        (anchorOffset === 0 || node.nodeType === TEXT_NODE)
```
I think this basically counts as zeugma. My code is poetry.

sebmarkbage commented on src/renderers/dom/shared/ReactDOMSelection.js:
```diff
@@ -39,59 +32,116 @@ function getModernOffsets(node) {
+  let node = outerNode;
+  let parentNode = null;
+
+  outer: while (true) {
```
I think V8 (at least the old one) deopts in some cases when it thinks it might be an infinite loop. We have more of these though so we can probably do a pass over all of them and work around it, if that's the case.

sebmarkbage commented on src/renderers/dom/shared/ReactDOMSelection.js:
```diff
@@ -39,59 +32,116 @@ function getModernOffsets(node) {
+  focusOffset,
+) {
+  let length = 0;
+  let start = null;
```
`-1` might help these be monomorphic in type-driven optimizations.

sophiebits replied:
k

Gregoirevda commented on src/renderers/dom/shared/ReactDOMSelection.js:
```diff
@@ -39,59 +32,116 @@ function getModernOffsets(node) {
+        length += node.nodeValue.length;
+      }
+
+      if ((next = node.firstChild) === null) {
```
`node.firstChild` can never be undefined?

sophiebits replied:
No, only null or an actual Node.

gaearon commented:
Let’s land it?

ojab commented on src/renderers/dom/shared/__tests__/ReactDOMSelection-test.js:
```diff
@@ -0,0 +1,206 @@
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
```
MIT MIT MIT

sophiebits replied:
Thank you. Will fix in the morning.

gaearon replied:
Maybe we need a lint rule 😄
